File: Medium/316. Remove Duplicate Letters/316. Remove Duplicate Letters.js
/**
 * @param {string} s
 * @return {string}
 */
var removeDuplicateLetters = function (s) {
    const stack = [];
    const lastOccurrence = new Array(26).fill(-1); // Initialize with -1 instead of 0.

    // Populate lastOccurrence with the last index of each character in s.
    for (let i = 0; i < s.length; i++) {
        lastOccurrence[s.charCodeAt(i) - 'a'.charCodeAt(0)] = i;
    }

    const seen = new Set(); // Use a set to track seen characters.

    for (let i = 0; i < s.length; i++) {
        const ch = s[i];

        if (!seen.has(ch)) {
            while (
                stack.length > 0 &&
                ch < stack[stack.length - 1] &&
                i <
                    lastOccurrence[
                        stack[stack.length - 1].charCodeAt(0) -
                            'a'.charCodeAt(0)
                    ]
            ) {
                const removedChar = stack.pop();
                seen.delete(removedChar);
            }

            seen.add(ch);
            stack.push(ch);
        }
    }

    return stack.join('');
};

File: Medium/316. Remove Duplicate Letters/316. Remove Duplicate Letters.java
class Solution {
    public String removeDuplicateLetters(String s) {
        StringBuilder stack = new StringBuilder();
        int[] lastOccurrence = new int[26];
        boolean[] seen = new boolean[26];

        // Populate lastOccurrence with the last index of each character in s.
        for (int i = 0; i < s.length(); i++) {
            lastOccurrence[s.charAt(i) - 'a'] = i;
        }

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            if (!seen[ch - 'a']) {
                while (stack.length() > 0 && ch < stack.charAt(stack.length() - 1)
                       && i < lastOccurrence[stack.charAt(stack.length() - 1) - 'a']) {
                    seen[stack.charAt(stack.length() - 1) - 'a'] = false;
                    stack.deleteCharAt(stack.length() - 1);
                }

                seen[ch - 'a'] = true;
                stack.append(ch);
            }
        }

        return stack.toString();
    }
}
